hypothesis-python/src/hypothesis/extra/gufunc.py

```diff
@@ -0,0 +1,451 @@
+        (``'int64'``), type (``np.int64``), or numpy `dtype`
+        (``np.dtype('int64')``). Built in Python types (`int`, `float`, etc)
+        also work. A single `dtype` can be supplied for all arguments.
+    elements : list-like of strategy
+        Strategies to fill in array elements on a per argument basis. One can
+        also specify a single strategy
+        (e.g., :func:`hypothesis.strategies.floats`)
+        and have it applied to all arguments.
+    unique : list-like of bool
+        Boolean flag to specify if all elements in an array must be unique.
+        One can also specify a single boolean to apply it to all arguments.
+
+    Returns
+    -------
+    res : tuple of ndarrays
+        Resulting ndarrays with shape of `shapes` and elements from `elements`.
+
+    """
+    if isinstance(shapes, SearchStrategy):
+        shapes = draw(shapes)
+    n = len(shapes)
+
+    # Need this since broadcast_to does not like vars of type type
+    if isinstance(dtype, type):
+        dtype = [dtype]
+    dtype = np.broadcast_to(dtype, (n,))
+
+    elements = np.broadcast_to(elements, (n,))
+    unique = np.broadcast_to(unique, (n,))
+
+    # This could somewhat easily be done using builds and avoid need for
+    # composite if shape is always given and not strategy. Otherwise, we need
+    # to chain strategies and probably not worth the effort.
+    res = tuple(draw(_arrays(dd, ss, elements=ee, unique=uu))
+                for dd, ss, ee, uu in zip(dtype, shapes, elements, unique))
+    return res
+
+
+def _signature_map(map_dict, parsed_sig):
+    """Map values found in parsed gufunc signature.
+
+    Parameters
+    ----------
+    map_dict : dict of str to int
+        Mapping from `str` dimension names to `int`. All strings in
+        `parsed_sig` must have entries in `map_dict`.
+    parsed_sig : list-like of tuples of str
+        gufunc signature that has already been parsed, e.g., using
+        `parse_gufunc_signature`.
+
+    Returns
+    -------
+    shapes : list of tuples of int
+        list of tuples where each tuple is the shape of an argument.
+
+    """
+    shapes = [tuple(map_dict[k] for k in arg) for arg in parsed_sig]
+    return shapes
+
+
+def _gufunc_arg_shapes(parsed_sig, min_side, max_side):
+    """Strategy to generate array shapes for arguments to a function consistent
+    with its signature.
+
+    Parameters
+    ----------
+    parsed_sig : list-like of tuples of str
+        gufunc signature that has already been parsed, e.g., using
+        `parse_gufunc_signature`.
+    min_side : defaultdict of str to int
+        Minimum size of any of the dimensions in `parsed_sig`.
+    max_side : defaultdict of str to int
+        Maximum size of any of the dimensions in `parsed_sig`.
+
+    Returns
+    -------
+    shapes : list of tuples of int
+        list of tuples where each tuple is the shape of an argument.
+
+    """
+    # Skipping validation on min and max sides since this function is private.
```

Comment:
For private functions, we typically use assertions for validation.  That way fuzzing-style tests can catch our internal usage errors, without adding branches that would have to be covered by more tests.

Reply:
so call `order_check_min_max(min_side, max_side)` again here?

Reply:
Or `assert min_side is None or max_side is None or min_side <= max_side`, yes.